airflow/models/xcom_arg.py

```diff
@@ -0,0 +1,96 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from typing import Any, Dict, List, Optional, Union
+
+from airflow.models.baseoperator import BaseOperator
+from airflow.models.xcom import XCOM_RETURN_KEY
+
+
+class XComArg:
+    """
+    Class that represents a XCom push from a previous operator.
+    Defaults to "return_value" as only key.
+    """
+
+    def __init__(self, operator: BaseOperator, keys: Optional[List[str]] = None):
+        self._operator = operator
+        self._keys = keys or [XCOM_RETURN_KEY]
+
+    def __eq__(self, other):
+        return self._operator == other._operator  # pylint: disable=protected-access
+
+    def __lshift__(self, other):
+        self.set_upstream(other)
+        return self
+
+    def __rshift__(self, other):
+        self.set_downstream(other)
+        return self
+
+    def set_upstream(self, task_or_task_list: Union[BaseOperator, List[BaseOperator]]):
+        """
+        Proxy to underlying operator set_upstream method
+        """
+        self._operator.set_upstream(task_or_task_list)
+
+    def set_downstream(
+        self, task_or_task_list: Union[BaseOperator, List[BaseOperator]]
+    ):
+        """
+        Proxy to underlying operator set_downstream method
+        """
+        self._operator.set_downstream(task_or_task_list)
```

Comment:
Reopening conversation regarding https://github.com/databand-ai/airflow/pull/3#discussion_r415021921 as force push is banning me from commenting further on that thread. 

Regarding the option you showed @turbaszek I believe it should be allowed, but not on the XComArg instance, rather the PythonFunctionalOperator/@task decorator. As defined in https://github.com/apache/airflow/issues/8057 the @task decorator should wrap the function into a task instance of the new `PythonFunctionalOperator` (defined in https://github.com/apache/airflow/issues/8056). This should allow for upstream setting in an Airflow fashion without mixing XComArg in the mix. Example with sensors:

```python
with DAG("example_dag"):
    # do something long running
    sensor = WaitForMagicSensor(task_id="sensor", ...)
    top = process_data()
 
    sensor >> process_data 
```

Reply:
I think we are raising another topic here:  how are you going to connect  `sensor` to the specific instance of process_data here? ( you can have multiple process_data in the same DAG, you can have process_data used by other Dags? ) 

>>, << functionality is isolated and doesn't affect other implementation, please see my comment below (summary) and let me know what you think.

Reply:
`process_data` should be an operator instance (so it can only be used once in a DAG). A way to reuse it is to add a `copy` function in the operator instance to create new instances of the functional operator: https://github.com/apache/airflow/issues/8056.

I know its a bit of a workaround. Not super happy about it either. But I'm worried that allowing a `task` to be reused multiple times in a DAG is changing the operator behaviour/Airflow DAG definition too much. 

How do we ensure unique task_id? In my proposal that is done thanks to Python enforcing uniqueness in function names. We can append a number at the end, but I'm afraid it will make it difficult to reason.

Reply:
1. we are changing a concept of airflow by saying that some code definition is connected to specific DAG. (You could define new operators without attaching them to DAG). I am more in favor saying that only f() creates the real operator and the f itself is just a factory of operators, same as PythonOperator, or any other Operator.
2. we need to add support for user defined task_id. A lot of airflow users like to have very readable names in the graph. In addition it possible to manage task_id assign policy on DAG level, so every new Operator has unique ID (a nice feature for airflow even decide functional operators)
3. how do you see this `copy` function?

Reply:
1. Agree on that. I'm worried that making it a factory wont be as intuitive, as the task instance is not accessible at all as __call__ returns the XComArg directly instead. Making it an operator that binds itself to the dag if needed on the __call__ method was my main proposed approach.
2. Task id is by default function name in this other approach.
3. `copy` method is a similar approach than dagster `alias` method: https://dagster.readthedocs.io/en/0.7.8/sections/tutorial/reusable.html

Reply:
1.a This is why making XArgCom behavior similar to task can make some sense ( at least at task wiring operations, it looks like this is the most the most frequently used feature on `task` object).  1.b you can also access task using result.operator .  1.c   1.a and 1.b is not the best thing to have, but changing standard pattern in order to support it can become a strong architecture problem.
2. agree. and it's easy to implement a mechanism to add _idx for all extra usages of the function in the DAG in case it's required. ( we have a working example at dbnd) 
3. dagster, prefect, dbnd, metaflow and all others usually will have a function as a factory that can be modified before the construction of the real Task/Operator/Solid/Node. I would keep this approach. .alias is a good name, and this can change default value of every next operator to be based on the new value (instead of default). (2) can help with multiple instances without .alias. 

4. in dbnd we also have some complication around making possible to define outputs via function parameters ( complicated mechanism) that makes defaul function call `f()` behaviour  non-consistent in some cases. we don't have this requirement now, so we can assume that the thing is returned  from f() is always have reference to task.  and whatever runs inside functional can always get  access to `self` of Operator via. current_context().task. (we can even create extra function that will do it directly.  current_operator())